### ISSUE TITLE
Chapter: Modules

### ISSUE THREAD

**Issue by rauschma** [rauschma/exploring-js#20]

**Comment by rightpad** [comment]
In section 22.1, underneath the code snippit reads:

> `myModule` is a global variable.

However the code snippit doesn't have a variable named `myModule`, did you mean `myLibrary`?

**Comment by mathiasbynens** [comment]
It might be useful to mention the `.mjs` extension, which is increasingly getting adopted by tooling as a signal that the file is a JavaScript module as opposed to a classic script.

The experimental modules implementation in Node.js only works for `.mjs`. Recently, the Node.js Modules Working Group agreed on the [“minimal kernel”](https://github.com/nodejs/modules/blob/master/doc/plan-for-new-modules-implementation.md) for what will one day become the unflagged implementation, and it states that `.mjs` will continue to be supported out-of-the-box. (There is a plan to allow other extensions as well at the cost of additional configuration.)

**Comment by rauschma** [comment]
@mathiasbynens The plan is to do both at the same time: Switch to `.mjs` (once it is supported without a flag by Node.js) and explain the rationale behind `.mjs`. With “Impatient JS”, I stick to established things.

**Comment by rauschma** [comment]
@da-moo Thanks! A fix will be in the next release.

**Comment by mcast** [comment]
First, thanks for compiling a useful set of explanations!

> export default foo();

Did you mean the `()` in that?

When I try it, I seem to be exporting the result of calling the function (which makes sense because `()` invoked it), but this doesn't fit  either ther apparent intent of the statement or the notion of exports being static.

```
ubuntu@pettle:~/tmp$ node --experimental-modules B.mjs 
(node:7610) ExperimentalWarning: The ESM module loader is experimental.
called foo
first line of B
bar = string result
ubuntu@pettle:~/tmp$ cat A.mjs 
function foo() {
  console.log("called foo");
  return "string result";
}
export default foo();
ubuntu@pettle:~/tmp$ cat B.mjs 
console.log("first line of B");
import bar from './A.mjs';
console.log("bar = " + bar);
ubuntu@pettle:~/tmp$ node --experimental-modules B.mjs 
(node:7634) ExperimentalWarning: The ESM module loader is experimental.
called foo
first line of B
bar = string result
ubuntu@pettle:~/tmp$ node -v
v8.11.4
```

Or would this describe the intent of the example?

> export default foo(); // export the return value of calling foo()

**Comment by rauschma** [comment]
@mcast Good point. Both do indeed make sense. I’ll try to make this clearer in the next release.

**Comment by ApolloTang** [comment]
Thank you for this book. I'm enjoying reading it! 

I have a question: 

RE:  section [https://exploringjs.com/impatient-js/ch_modules.html#naming-modules](https://exploringjs.com/impatient-js/ch_modules.html#naming-modules )

![image](https://user-images.githubusercontent.com/1150961/174403945-77ff52a5-6c10-48b6-a3c7-892734432946.png)

could you please explain why underscore-casing does not work for default import and why it is not a good choice for functions? 

Thanks!

**Comment by docwarems** [comment]
Axel, this article (or chapter) made by day.

I've been a Java programmer almost from the Java beginnings, who had to move to TS/node a year ago and work on a node project which have been developed for years.
The different module and import/export paradigms from JS, TS, node, nest drove me crazy. And if you work through tutorials on each of these technics you get always only the module view from this technics perspective. In the project I'm working all these technics are mixed and I haven't got a true view on how the module system works even for one year. Consequently refactorings habe been a big pain because I usually needed most of the time with updating dependencies by trial & error because there was no true understanding.

Your article is the first source I found where everything is put together in a compact way starting with the history and how it evolved.

Now I understand my project a lot better. Thank you very much!

**Comment by RaushniGupta** [comment]
Hello there, very Interesting! It's a wonderful blog. It's an excellent work of writing. Thank you for taking the time to write and publish such an important piece. Thanks buddy! Visit Us: [Delhi Tourism Blog](https://delhitourismgov.blogspot.com/)

**Comment by vovochka-dev** [comment]
Thank you, for your work. I have question about default export.
> // Default exports
>export default function f() {} // declaration with optional name

and 

> 27.7.1 The two styles of default-exporting 
There are two styles of doing default exports.

>First, we can label existing declarations with export default:
```
export default function myFunc() {} // no semicolon!
export default class MyClass {} // no semicolon!
```

I tried to figure out what existing declarations really mean, maybe it's about hoisting thing. But then I realized that this is actually named function expression and named class expression.

After `export default` have to be expression, not declaration (or existing expression).
ESM creators have even considered another more explicit syntax with '=': `export default = function myFunc() {};`
Under the hood implicitly we set value (expression) to `default` variable. 

```
const funcExpression = function myFunc() {}
export default myFuncExpression;

const classExpression = class MyClass {}
export default classExpression;
```
So, is it existing declarations or expressions?

**Comment by rauschma** [comment]
> I tried to figure out what existing declarations really mean, maybe it's about hoisting thing. But then I realized that this is actually named function expression and named class expression.

@vovochka-dev: I should probably remove the word “existing”. I agree that that word makes this sentence slightly confusing.

They really are declarations – check out [the three grammar rules for `export default`](https://tc39.es/ecma262/#sec-scripts-and-modules) in the ECMAScript specification. Two of them are for declarations.

**Comment by rauschma** [comment]
@ApolloTang:

> could you please explain why underscore-casing does not work for default import and why it is not a good choice for functions?

The common practice for JavaScript is to use camel case for functions. I’m going back and forth on underscore case: Maybe it shouldn’t be used at all (except for constants).

**Comment by vovochka-dev** [comment]
> > I tried to figure out what existing declarations really mean, maybe it's about hoisting thing. But then I realized that this is actually named function expression and named class expression.
> 
> @vovochka-dev: I should probably remove the word “existing”. I agree that that word makes this sentence slightly confusing.
> 
> They really are declarations – check out [the three grammar rules for `export default`](https://tc39.es/ecma262/#sec-scripts-and-modules) in the ECMAScript specification. Two of them are for declarations.

Thank you, Indeed it's declaration:

```
export default function add(a, b) {return a + b;}
console.log(add(2,2)) //4
```
One more thing. When I understood it was "aha moment" about `export default const` problem.  Consider some changes: 
>// Not legal JavaScript!
>export default const foo = 1, bar = 2, baz = 3;
+// Legal JavaScript
+export const foo = 1, bar = 2, baz = 3;

And also Cheat sheet does not contain this syntax:
>export function f() {}
export const one = 1;
export {foo, b as bar};
+export const foo = 1, bar = 2, baz = 3;

**Comment by yanjankaf** [comment]
I'm having trouble understanding this part,?

> On the plus side, AMD modules can be executed directly. In contrast, CommonJS modules must either be compiled before deployment or custom source code must be generated and evaluated dynamically [(think eval())](https://exploringjs.com/impatient-js/ch_dynamic-code-evaluation.html#eval). That isn’t always permitted on the web.

------
I think the author is trying to say something like this 

Since you have a script element you can load as many scripts as you want dynamically, but can't you do the same in common JS? check some conditions, then call `require`?

So basically since the common.js modules are not supported by javascript by default, you need to compile this to javascript, but why would you do this? if the node supports the `require` function let it use it. well ofcourse if you are putting the code you have written using require, you need some sort of polyfill for that, 

I have no idea about  this sentence -  "Custom source code must be generated and evaluated dynamically  [(think eval())](https://exploringjs.com/impatient-js/ch_dynamic-code-evaluation.html#eval)"

**Comment by MegaManSec** [comment]
The ahref for `[29.14 Loading modules dynamically via import() [ES2020] (advanced)](https://exploringjs.com/js/book/ch_modules.html#dynamic-imports)` links to `[29.1.2 Dynamic imports](https://exploringjs.com/js/book/ch_modules.html#dynamic-imports)`

**Comment by waldoluisribeiro** [comment]
[29.2 JavaScript source code formats](https://exploringjs.com/js/book/ch_modules.html#javascript-source-code-formats)

- ES6 brought built-in modules, but the source code formats that came before **them, are** still around, too. --> ES6 brought built-in modules, but the source code formats that came before **them are** still around, too.

[29.2.1 Code before built-in modules was written in ECMAScript 5](https://exploringjs.com/js/book/ch_modules.html#code-before-built-in-modules-was-written-in-ecmascript-5)

- Before we get to built-in modules (which were introduced with ES6), all code that we’ll **see, will** be written in ES5. --> Before we get to built-in modules (which were introduced with ES6), all code that we’ll **see will** be written in ES5.

[29.3 Before we had modules, we had scripts](https://exploringjs.com/js/book/ch_modules.html#scripts)

- This way of simulating **modules, has** several issues --> This way of simulating **modules has** several issues
